flink-runtime/src/test/java/org/apache/flink/runtime/state/IncrementalRemoteKeyedStateHandleTest.java

```diff
@@ -243,6 +244,60 @@ public void testNonEmptyIntersection() {
         assertEquals(handle.getStateHandleId(), newHandle.getStateHandleId());
     }
 
+    @Test
+    public void testConcurrentCheckpointSharedStateRegistration() throws Exception {
+        StateHandleID handleID = new StateHandleID("1.sst");
+        StreamStateHandle streamHandle1 = new ByteStreamStateHandle("file-1", new byte[] {'s'});
+        StreamStateHandle streamHandle2 = new ByteStreamStateHandle("file-2", new byte[] {'s'});
+
+        SharedStateRegistry registry = new SharedStateRegistryImpl();
+
+        UUID backendID = UUID.randomUUID();
+
+        IncrementalRemoteKeyedStateHandle handle1 =
+                new IncrementalRemoteKeyedStateHandle(
+                        backendID,
+                        KeyGroupRange.of(0, 0),
+                        1L,
+                        placeSpies(
+                                new HashMap<StateHandleID, StreamStateHandle>() {
+                                    {
+                                        put(handleID, streamHandle1);
+                                    }
+                                }),
```

Comment:
`Collections.singletonMap`?

ditto: line 279

Reply:
Collections.singletonMap return a immutable map, but we need replace the value by IncrementalRemoteKeyedStateHandleTest#placeSpies .

Reply:
Fair enough.